File: turms-client-dart/lib/src/driver/service/message_service.dart
import 'dart:async';
import 'dart:math';

import '../../exception/response_exception.dart';
import '../../extension/int_extensions.dart';
import '../../model/proto/notification/turms_notification.pb.dart';
import '../../model/proto/request/turms_request.pb.dart';
import '../../model/response_status_code.dart';
import '../state_store.dart';
import 'base_service.dart';

typedef NotificationListener = void Function(TurmsNotification notification);

class TurmsRequestContext {
  final Completer<TurmsNotification> completer;
  final Timer? timeoutTimer;

  TurmsRequestContext(this.completer, this.timeoutTimer);
}

class DriverMessageService extends BaseService {
  static const int randomMax = 1 << 32;

  final Random _random = Random();

  late final int _requestTimeoutMillis;
  late final int _minRequestIntervalMillis;
  final List<NotificationListener> _notificationListeners = [];
  final Map<int, TurmsRequestContext> _idToRequest = {};

  DriverMessageService(StateStore stateStore, int? requestTimeoutMillis,
      int? minRequestIntervalMillis)
      : super(stateStore) {
    _requestTimeoutMillis =
        requestTimeoutMillis == null || requestTimeoutMillis <= 0
            ? 60 * 1000
            : requestTimeoutMillis;
    _minRequestIntervalMillis = minRequestIntervalMillis ?? 0;
  }

  // Listeners

  void addNotificationListener(NotificationListener listener) =>
      _notificationListeners.add(listener);

  void removeNotificationListener(NotificationListener listener) =>
      _notificationListeners.remove(listener);

  void _notifyNotificationListeners(TurmsNotification notification) {
    for (final listener in _notificationListeners) {
      try {
        listener.call(notification);
      } catch (e, s) {
        print('$e\n$s');
      }
    }
  }

  // Request and notification

  Future<TurmsNotification> sendRequest(TurmsRequest request) async {
    if (request.hasCreateSessionRequest()) {
      if (stateStore.isSessionOpen) {
        throw ResponseException(
            code: ResponseStatusCode.clientSessionAlreadyEstablished);
      }
    } else if (!stateStore.isConnected || !stateStore.isSessionOpen) {
      throw ResponseException(
          code: ResponseStatusCode.clientSessionHasBeenClosed);
    }
    final now = DateTime.now().millisecondsSinceEpoch;
    final difference = now - stateStore.lastRequestDate;
    final isFrequent = _minRequestIntervalMillis > 0 &&
        difference <= _minRequestIntervalMillis;
    if (isFrequent) {
      throw ResponseException(
          code: ResponseStatusCode.clientRequestsTooFrequent);
    }
    final requestId = _generateRandomId();
    request.requestId = requestId.toInt64();
    final payload = request.writeToBuffer();
    stateStore.tcp!.writeVarIntLengthAndBytes(payload);
    final timeoutTimer = _requestTimeoutMillis > 0
        ? Timer(Duration(milliseconds: _requestTimeoutMillis), () {
            final context = _idToRequest.remove(requestId);
            context?.completer.completeError(
                ResponseException(code: ResponseStatusCode.requestTimeout));
          })
        : null;
    final completer = Completer<TurmsNotification>();
    final requestContext = TurmsRequestContext(completer, timeoutTimer);
    _idToRequest[requestId] = requestContext;
    stateStore.lastRequestDate = now;
    return completer.future;
  }

  void didReceiveNotification(TurmsNotification notification) {
    final isResponse =
        !notification.hasRelayedRequest() && notification.hasRequestId();
    if (isResponse) {
      final requestId = notification.requestId.toInt();
      final context = _idToRequest.remove(requestId);
      if (context != null) {
        context.timeoutTimer?.cancel();
        if (notification.hasCode()) {
          if (ResponseStatusCode.isSuccessCode(notification.code)) {
            context.completer.complete(notification);
          } else {
            context.completer.completeError(
                ResponseException.fromNotification(notification));
          }
        } else {
          context.completer.completeError(ResponseException(
              requestId:
                  notification.hasRequestId() ? notification.requestId : null,
              code: ResponseStatusCode.invalidNotification,
              reason: 'The code is missing'));
        }
      }
    }
    _notifyNotificationListeners(notification);
  }

  int _generateRandomId() {
    int id;
    do {
      id = _random.nextInt(randomMax);
    } while (_idToRequest.containsKey(id));
    return id;
  }

  void _rejectRequestCompleter(ResponseException exception) {
    _idToRequest.removeWhere((key, context) {
      context.completer.completeError(exception);
      return true;
    });
  }

  @override
  Future<void> close() async {
    onDisconnected();
    return;
  }

  @override
  void onDisconnected({Object? error, StackTrace? stackTrace}) {
    final exception = ResponseException(
        code: ResponseStatusCode.clientSessionHasBeenClosed,
        cause: error,
        stackTrace: stackTrace);
    _rejectRequestCompleter(exception);
  }
}

File: turms-client-cpp/src/turms/client/driver/service/message_service.cpp
#include "turms/client/driver/service/message_service.h"

namespace turms {
namespace client {
namespace driver {
namespace service {
MessageService::MessageService(boost::asio::io_context& ioContext,
                               StateStore& stateStore,
                               const boost::optional<int>& requestTimeout,
                               const boost::optional<int>& minRequestInterval)
    : BaseService(ioContext, stateStore),
      requestTimeout_(std::chrono::milliseconds{requestTimeout.get_value_or(60 * 1000)}),
      enableRequestTimeout_(requestTimeout_.count() > 0),
      minRequestInterval_(minRequestInterval.get_value_or(0)) {
}

auto MessageService::sendRequest(model::proto::TurmsRequest& request)
    -> boost::future<TurmsNotification> {
    if (request.has_create_session_request()) {
        if (stateStore_.isSessionOpen) {
            return boost::make_exceptional_future<TurmsNotification>(
                ResponseException{model::ResponseStatusCode::kClientSessionAlreadyEstablished});
        }
    } else if (!stateStore_.isConnected || !stateStore_.isSessionOpen) {
        return boost::make_exceptional_future<TurmsNotification>(
            ResponseException{model::ResponseStatusCode::kClientSessionHasBeenClosed});
    }

    auto now = std::chrono::system_clock::now().time_since_epoch().count();
    auto difference = now - stateStore_.lastRequestDate;
    auto isFrequent = minRequestInterval_ > 0 && difference <= minRequestInterval_;
    if (isFrequent) {
        return boost::make_exceptional_future<TurmsNotification>(
            ResponseException{model::ResponseStatusCode::kClientRequestsTooFrequent});
    }

    while (true) {
        auto requestId = generateRandomId();
        // Use "try_emplace" when upgrading to C++17
        auto result = idToRequest_.emplace(
            requestId, TurmsRequestContext{boost::promise<TurmsNotification>(), boost::none});
        if (!result.second) {
            continue;
        }
        TurmsRequestContext& context = result.first->second;
        boost::promise<TurmsNotification>& promise = context.promise;
        stateStore_.lastRequestDate = now;
        try {
            request.set_request_id(requestId);
            auto payload = std::make_shared<std::vector<uint8_t>>(request.ByteSizeLong());
            request.SerializeToArray(payload->data(), payload->size());
            stateStore_.tcp->writeVarIntLengthAndBytes(payload);
        } catch (const std::exception& e) {
            idToRequest_.erase(requestId);
            promise.set_exception(e);
            return promise.get_future();
        }

        if (enableRequestTimeout_) {
            boost::asio::steady_timer timer{ioContext_, requestTimeout_};
            timer.async_wait([weakThis = std::weak_ptr<MessageService>(shared_from_this()),
                              requestId](const boost::system::error_code& e) {
                if (e == boost::asio::error::operation_aborted) {
                    return;
                }
                auto sharedThis = weakThis.lock();
                if (sharedThis == nullptr) {
                    return;
                }
                auto& idToRequest = sharedThis->idToRequest_;
                auto it = idToRequest.find(requestId);
                if (it == idToRequest.end()) {
                    return;
                }
                it->second.promise.set_exception(
                    ResponseException{model::ResponseStatusCode::kRequestTimeout});
                idToRequest.erase(it);
            });
            context.timeoutTimer = std::move(timer);
        }
        return promise.get_future();
    }
}

auto MessageService::didReceiveNotification(const MessageService::TurmsNotification& notification)
    -> void {
    const bool isResponse = !notification.has_relayed_request() && notification.has_request_id();
    if (isResponse) {
        auto it = idToRequest_.find(notification.request_id());
        if (it != idToRequest_.end()) {
            auto& context = it->second;
            auto& promise = context.promise;
            if (notification.has_code()) {
                auto& timer = context.timeoutTimer;
                if (timer) {
                    timer->cancel();
                }
                if (model::notification::isSuccess(notification)) {
                    promise.set_value(notification);
                } else {
                    promise.set_exception(ResponseException{notification});
                }
            } else {
                promise.set_exception(ResponseException{
                    model::ResponseStatusCode::kInvalidNotification, "The code is missing"});
            }
            idToRequest_.erase(it);
        }
    }
    notifyNotificationListeners(notification);
}

auto MessageService::close() -> boost::future<void> {
    onDisconnected(boost::none);
    return boost::make_ready_future();
}

auto MessageService::onDisconnected(const boost::optional<std::exception>& exception) -> void {
    rejectRequestPromises(exception::ResponseException{
        model::ResponseStatusCode::kClientSessionHasBeenClosed, "", exception});
}

auto MessageService::generateRandomId() const -> int64_t {
    int64_t id;
    do {
        id = random::nextPositiveInt64();
    } while (idToRequest_.count(id) > 0);
    return id;
}

auto MessageService::notifyNotificationListeners(
    const MessageService::TurmsNotification& notification) const -> void {
    for (const auto& listener : notificationListeners_) {
        listener(notification);
    }
}

auto MessageService::rejectRequestPromises(const std::exception& exception) -> void {
    for (auto it = idToRequest_.begin(); it != idToRequest_.end();) {
        it = idToRequest_.erase(it);
        TurmsRequestContext& context = it->second;
        context.promise.set_exception(exception);
        if (context.timeoutTimer) {
            context.timeoutTimer->cancel();
        }
    }
}
}  // namespace service
}  // namespace driver
}  // namespace client
}  // namespace turms